huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go

```diff
@@ -0,0 +1,124 @@
+package smn
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/chnsz/golangsdk/openstack/common/tags"
+	"github.com/chnsz/golangsdk/openstack/smn/v2/topics"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+func DataSourceTopic() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceTopicRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"topic_urn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"display_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"enterprise_project_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"tags": {
+				Type:     schema.TypeMap,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"push_policy": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceTopicRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.SmnV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating SMN client: %s", err)
+	}
+
+	allTopics, err := topics.List(client).Extract()
+	if err != nil {
+		return diag.Errorf("unable to list topics: %s ", err)
+	}
+
+	log.Printf("[DEBUG] Retrieved SMN topics: %#v", allTopics)
+
+	filter := map[string]interface{}{
+		"TopicUrn":            d.Get("topic_urn"),
+		"Name":                d.Get("name"),
+		"EnterpriseProjectId": d.Get("enterprise_project_id"),
+		"DisplayName":         d.Get("display_name"),
+	}
+
+	filterTopics, err := utils.FilterSliceWithField(allTopics, filter)
+	if err != nil {
+		return diag.Errorf("filter topics failed: %s", err)
+	}
+
+	if len(filterTopics) < 1 {
+		return diag.Errorf("Your query returned no results. " +
+			"Please change your search criteria and try again.")
+	}
+
+	if len(filterTopics) > 1 {
+		return diag.Errorf("Your query returned more than one result." +
+			" Please try a more specific search criteria")
+	}
```

Comment:
1. pay attention to the case (**your**).
2. total two sentences and other codes no more than 120 characters, can be written on a single line.

Reply:
deleted